File: solutions/typescript/rational-numbers/1/rational-numbers.ts
export class Rational {
  numerator: number
  denominator: number

  constructor(numerator: number, denominator: number) {
    this.numerator = numerator
    this.denominator = denominator
  }

  add(other: Rational) {
    const numerator = this.numerator * other.denominator + other.numerator * this.denominator
    const denominator = this.denominator * other.denominator
    if (numerator === 0) {
        return new Rational(0, 1)
    }
    return new Rational(numerator, denominator)
  }

  sub(other: Rational) {
    const numerator = this.numerator * other.denominator - other.numerator * this.denominator
    const denominator = this.denominator * other.denominator
    if (numerator === 0) {
        return new Rational(0, 1)
    }
    return new Rational(numerator, denominator)
  }

  mul(other: Rational) {
      const numerator = this.numerator * other.numerator
      const denominator = this.denominator * other.denominator
      if (numerator === 0) {
          return new Rational(0, 1)
      }
      if (denominator < 0) {
          return new Rational(-numerator, -denominator)
      }
      return new Rational(numerator, denominator).reduce()
  }

  div(other: Rational) {
      const numerator = this.numerator * other.denominator
      const denominator = this.denominator * other.numerator
      return new Rational(numerator, denominator).reduce()
  }

  abs() {
    const numerator = Math.abs(this.numerator)
    const denominator = Math.abs(this.denominator)
    return new Rational(numerator, denominator).reduce()
  }

  exprational(exp: number) {
    if (exp === 0) {
        return new Rational(1, 1)
    }
    if (exp < 0) {
        const numerator = Math.pow(this.denominator, Math.abs(exp))
        const denominator = Math.pow(this.numerator, Math.abs(exp))
        if (denominator < 0) {
            return new Rational(-numerator, -denominator)
        }
        return new Rational(numerator, denominator)
    }
    const numerator = Math.pow(this.numerator, exp)
    const denominator = Math.pow(this.denominator, exp)
    if (numerator === 0) {
        return new Rational(0, 1)
    }
    if (denominator < 0) {
        return new Rational(-numerator, -denominator)
    }
    return new Rational(numerator, denominator)
  }

  expreal(exp: number) {
      // raise the real number to the power of the rational number
      const numerator = Math.pow(exp, this.numerator)
      return Math.pow(numerator, 1 / this.denominator)
  }

  reduce() {
      const gcd = (a: number, b: number): number => b === 0 ? a: gcd(b, a % b)
      const divisor = gcd(this.numerator, this.denominator)
      const numerator = this.numerator / divisor
      const denominator = this.denominator / divisor
      if (denominator < 0) {
          return new Rational(-numerator, -denominator)
      }
      if (numerator === 0) {
          return new Rational(0, 1)
      }

      return new Rational(numerator, denominator)

  }
}

File: solutions/go/rational-numbers/1/rational_numbers.go
package rationalnumbers

import "math"

type Rational struct {
	numerator, denominator int
}

// Reduce simplifies a Rational, eg changing Rational{4, 2} into Rational{2, 1}.
func (r Rational) Reduce() Rational {
	gcd := gcd(abs(r.numerator), abs(r.denominator))
	if r.denominator < 0  && r.numerator < 0 {
		return Rational{abs(r.numerator / gcd), abs(r.denominator / gcd)}
	}
	if r.denominator < 0 {
		return Rational{-abs(r.numerator / gcd), abs(r.denominator / gcd)}
	}
	return Rational{r.numerator / gcd, r.denominator / gcd}
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func (r Rational) Add(s Rational) Rational {
	return Rational{r.numerator * s.denominator + s.numerator * r.denominator, r.denominator * s.denominator}.Reduce()
}

func (r Rational) Sub(s Rational) Rational {
	return Rational{r.numerator * s.denominator - s.numerator * r.denominator, r.denominator * s.denominator}.Reduce()
}

func (r Rational) Mul(s Rational) Rational {
	return Rational{r.numerator * s.numerator, r.denominator * s.denominator}.Reduce()
}

func (r Rational) Div(s Rational) Rational {
	return Rational{r.numerator * s.denominator, r.denominator * s.numerator}.Reduce()
}

func (r Rational) Abs() Rational {
	return Rational{abs(r.numerator), abs(r.denominator)}.Reduce()
}
func abs(a int) int {
	if a < 0 {
		return -a
	}
	return a
}

// Compute r ^ power, a rational raised to an int exponent.
func (r Rational) Exprational(power int) Rational {
	if power < 0 {
		return Rational{int(math.Pow(float64(r.denominator), float64(-power))), int(math.Pow(float64(r.numerator), float64(-power)))}.Reduce()
	}
	if power == 0 {
		return Rational{1, 1}
	}
	return Rational{int(math.Pow(float64(r.numerator), float64(power))), int(math.Pow(float64(r.denominator), float64(power)))}.Reduce()
}

// Compute base ^ r, an int raised to a rational.
func (r Rational) Expreal(base int) float64 {
	if r.denominator == 0 {
		return 0
	}
	if r.numerator == 0 {
		return 1
	}
	if r.denominator < 0 {
		return math.Pow(float64(base), float64(-r.numerator)/float64(-r.denominator))
	}
	return math.Pow(float64(base), float64(r.numerator)/float64(r.denominator))
}
